In [56]:
import pandas as pd 

df = pd.read_csv("finance_early_loan_repayment_prediction_dataset.xls")

df.head()

,Loan_ID,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type,Early_Repayment
0,1,68.0,Part-time,211300.0,628.0,116286,48,14.21,2199,118122.0,2,Yes,Auto,No
1,2,66.0,Full-time,99549.0,707.0,58967,24,13.38,2218,15269.0,3,No,Auto,No
2,3,31.0,Full-time,184931.0,664.0,90631,60,7.47,362,88882.0,4,No,Education,No
3,4,59.0,Full-time,124434.0,585.0,44255,36,6.22,1467,102787.0,4,No,Business,Yes
4,5,72.0,Full-time,108950.0,612.0,48376,60,4.93,990,64234.0,4,Yes,Business,Yes


In [57]:
print (df.shape)

print (df.columns)

print (df.info())

print (df.dtypes)

(387, 14)
Index(['Loan_ID', 'Age', 'Employment_Status', 'Annual_Income', 'Credit_Score',
       'Loan_Amount', 'Loan_Term_Months', 'Interest_Rate', 'Monthly_Payment',
       'Savings_Balance', 'Missed_Payments', 'Automatic_Payment', 'Loan_Type',
       'Early_Repayment'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 387 entries, 0 to 386
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            387 non-null    int64  
 1   Age                380 non-null    float64
 2   Employment_Status  380 non-null    object 
 3   Annual_Income      380 non-null    float64
 4   Credit_Score       380 non-null    float64
 5   Loan_Amount        387 non-null    int64  
 6   Loan_Term_Months   387 non-null    int64  
 7   Interest_Rate      387 non-null    float64
 8   Monthly_Payment    387 non-null    int64  
 9   Savings_Balance    379 non-null    float64
 10  Missed_Payments    

In [58]:
business_problem = "A financial institution wants to predict which customers are likely to repay their loans early."

ml_problem = "This is a classification problem because we want to predict whether the customer will repay the loan early or not."

prediction_goal = "Predict whether a customer will repay their loan early: Yes or No."

print("Business Problem:")
print(business_problem)

print("\nMachine Learning Problem:")
print(ml_problem)

print("\nPrediction Goal:")
print(prediction_goal)

Business Problem:
A financial institution wants to predict which customers are likely to repay their loans early.

Machine Learning Problem:
This is a classification problem because we want to predict whether the customer will repay the loan early or not.

Prediction Goal:
Predict whether a customer will repay their loan early: Yes or No.


In [59]:
feature_columns = [
"Age",
"Employment_Status",
"Annual_Income",
"Credit_Score",
"Loan_Amount",
"Loan_Term_Months",
"Interest_Rate",
"Monthly_Payment",
"Savings_Balance",
"Missed_Payments",
"Automatic_Payment",
"Loan_Type"
]

X = df[feature_columns]

y = df["Early_Repayment"]

print("Features (X):")
display(X.head())

print("Target (y):")
display(y.head())

Features (X):


,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type
0,68.0,Part-time,211300.0,628.0,116286,48,14.21,2199,118122.0,2,Yes,Auto
1,66.0,Full-time,99549.0,707.0,58967,24,13.38,2218,15269.0,3,No,Auto
2,31.0,Full-time,184931.0,664.0,90631,60,7.47,362,88882.0,4,No,Education
3,59.0,Full-time,124434.0,585.0,44255,36,6.22,1467,102787.0,4,No,Business
4,72.0,Full-time,108950.0,612.0,48376,60,4.93,990,64234.0,4,Yes,Business


Target (y):


0     No
1     No
2     No
3    Yes
4    Yes
Name: Early_Repayment, dtype: object

In [60]:
print("Shape of X:", X.shape)

print("Shape of y:", y.shape)

Shape of X: (387, 12)
Shape of y: (387,)


In [61]:
df_clean = df.copy()

print("Missing values before cleaning:")
print(df_clean.isnull().sum())

print("\nNumber of duplicate rows before cleaning:")
print(df_clean.duplicated().sum())

print("\nData types before cleaning:")
print(df_clean.dtypes)

Missing values before cleaning:
Loan_ID              0
Age                  7
Employment_Status    7
Annual_Income        7
Credit_Score         7
Loan_Amount          0
Loan_Term_Months     0
Interest_Rate        0
Monthly_Payment      0
Savings_Balance      8
Missed_Payments      0
Automatic_Payment    0
Loan_Type            8
Early_Repayment      0
dtype: int64

Number of duplicate rows before cleaning:
7

Data types before cleaning:
Loan_ID                int64
Age                  float64
Employment_Status     object
Annual_Income        float64
Credit_Score         float64
Loan_Amount            int64
Loan_Term_Months       int64
Interest_Rate        float64
Monthly_Payment        int64
Savings_Balance      float64
Missed_Payments        int64
Automatic_Payment     object
Loan_Type             object
Early_Repayment       object
dtype: object


In [62]:
df_clean["Age"] = df_clean["Age"].fillna(df_clean["Age"].median())
df_clean["Annual_Income"] = df_clean["Annual_Income"].fillna(df_clean["Annual_Income"].median())
df_clean["Credit_Score"] = df_clean["Credit_Score"].fillna(df_clean["Credit_Score"].median())
df_clean["Savings_Balance"] = df_clean["Savings_Balance"].fillna(df_clean["Savings_Balance"].median())

df_clean["Employment_Status"] = df_clean["Employment_Status"].fillna(df_clean["Employment_Status"].mode()[0])
df_clean["Loan_Type"] = df_clean["Loan_Type"].fillna(df_clean["Loan_Type"].mode()[0])

df_clean = df_clean.drop_duplicates()

df_clean.head()

,Loan_ID,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type,Early_Repayment
0,1,68.0,Part-time,211300.0,628.0,116286,48,14.21,2199,118122.0,2,Yes,Auto,No
1,2,66.0,Full-time,99549.0,707.0,58967,24,13.38,2218,15269.0,3,No,Auto,No
2,3,31.0,Full-time,184931.0,664.0,90631,60,7.47,362,88882.0,4,No,Education,No
3,4,59.0,Full-time,124434.0,585.0,44255,36,6.22,1467,102787.0,4,No,Business,Yes
4,5,72.0,Full-time,108950.0,612.0,48376,60,4.93,990,64234.0,4,Yes,Business,Yes


In [63]:
df_clean["Loan_ID"] = df_clean["Loan_ID"].astype(int)

df_clean["Age"] = df_clean["Age"].astype(int)
df_clean["Annual_Income"] = df_clean["Annual_Income"].astype(float)
df_clean["Credit_Score"] = df_clean["Credit_Score"].astype(int)
df_clean["Loan_Amount"] = df_clean["Loan_Amount"].astype(int)
df_clean["Loan_Term_Months"] = df_clean["Loan_Term_Months"].astype(int)
df_clean["Interest_Rate"] = df_clean["Interest_Rate"].astype(float)
df_clean["Monthly_Payment"] = df_clean["Monthly_Payment"].astype(int)
df_clean["Savings_Balance"] = df_clean["Savings_Balance"].astype(float)
df_clean["Missed_Payments"] = df_clean["Missed_Payments"].astype(int)

df_clean["Employment_Status"] = df_clean["Employment_Status"].astype("category")
df_clean["Automatic_Payment"] = df_clean["Automatic_Payment"].astype("category")
df_clean["Loan_Type"] = df_clean["Loan_Type"].astype("category")
df_clean["Early_Repayment"] = df_clean["Early_Repayment"].astype("category")

print("Data types after cleaning:")
print(df_clean.dtypes)

Data types after cleaning:
Loan_ID                 int64
Age                     int64
Employment_Status    category
Annual_Income         float64
Credit_Score            int64
Loan_Amount             int64
Loan_Term_Months        int64
Interest_Rate         float64
Monthly_Payment         int64
Savings_Balance       float64
Missed_Payments         int64
Automatic_Payment    category
Loan_Type            category
Early_Repayment      category
dtype: object


In [64]:
print("Missing values after cleaning:")
print(df_clean.isnull().sum())

print("\nNumber of duplicate rows after cleaning:")
print(df_clean.duplicated().sum())

print("\nShape of cleaned dataset:")
print(df_clean.shape)

df_clean.head()

Missing values after cleaning:
Loan_ID              0
Age                  0
Employment_Status    0
Annual_Income        0
Credit_Score         0
Loan_Amount          0
Loan_Term_Months     0
Interest_Rate        0
Monthly_Payment      0
Savings_Balance      0
Missed_Payments      0
Automatic_Payment    0
Loan_Type            0
Early_Repayment      0
dtype: int64

Number of duplicate rows after cleaning:
0

Shape of cleaned dataset:
(380, 14)


,Loan_ID,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type,Early_Repayment
0,1,68,Part-time,211300.0,628,116286,48,14.21,2199,118122.0,2,Yes,Auto,No
1,2,66,Full-time,99549.0,707,58967,24,13.38,2218,15269.0,3,No,Auto,No
2,3,31,Full-time,184931.0,664,90631,60,7.47,362,88882.0,4,No,Education,No
3,4,59,Full-time,124434.0,585,44255,36,6.22,1467,102787.0,4,No,Business,Yes
4,5,72,Full-time,108950.0,612,48376,60,4.93,990,64234.0,4,Yes,Business,Yes


In [65]:
df = df_clean.copy()    

In [66]:
from sklearn.compose import ColumnTransformer

from sklearn.preprocessing import OneHotEncoder, StandardScaler

In [67]:
X = df[feature_columns]

y = df["Early_Repayment"]

In [68]:
y = y.map({"No": 0, "Yes": 1})

In [69]:
numerical_features = [
"Age",
"Annual_Income",
"Credit_Score",
"Loan_Amount",
"Loan_Term_Months",
"Interest_Rate",
"Monthly_Payment",
"Savings_Balance",
"Missed_Payments"
]

In [70]:
categorical_features = [
"Employment_Status",
"Automatic_Payment",
"Loan_Type"
]

In [71]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numerical_features),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ]
)

In [72]:
print("Features before preprocessing:")

display(X.head())

Features before preprocessing:


,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type
0,68,Part-time,211300.0,628,116286,48,14.21,2199,118122.0,2,Yes,Auto
1,66,Full-time,99549.0,707,58967,24,13.38,2218,15269.0,3,No,Auto
2,31,Full-time,184931.0,664,90631,60,7.47,362,88882.0,4,No,Education
3,59,Full-time,124434.0,585,44255,36,6.22,1467,102787.0,4,No,Business
4,72,Full-time,108950.0,612,48376,60,4.93,990,64234.0,4,Yes,Business


In [73]:
print("Target after converting Yes/No to 1/0:")

display(y.head())

Target after converting Yes/No to 1/0:


0    0
1    0
2    0
3    1
4    1
Name: Early_Repayment, dtype: category
Categories (2, int64): [0, 1]

In [74]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
X,
y,
test_size=0.2,
random_state=42,
stratify=y
)

In [75]:
print("X_train shape:", X_train.shape)

print("X_test shape:", X_test.shape)

print("y_train shape:", y_train.shape)

print("y_test shape:", y_test.shape)

X_train shape: (304, 12)
X_test shape: (76, 12)
y_train shape: (304,)
y_test shape: (76,)


In [76]:
print("Training features:")
display(X_train.head())

print("Testing features:")
display(X_test.head())

Training features:


,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type
57,56,Full-time,129116.0,797,99838,36,16.31,2216,48297.0,1,Yes,Auto
315,34,Part-time,69945.0,468,47767,36,7.38,2889,109650.0,1,Yes,Business
202,34,Full-time,135564.0,721,62133,48,14.30,289,57224.0,3,Yes,Auto
164,32,Self-employed,212415.0,714,70840,12,5.87,2915,12893.0,1,Yes,Auto
77,41,Unemployed,125534.5,494,69694,48,12.79,2941,4064.0,2,Yes,Education


Testing features:


,Age,Employment_Status,Annual_Income,Credit_Score,Loan_Amount,Loan_Term_Months,Interest_Rate,Monthly_Payment,Savings_Balance,Missed_Payments,Automatic_Payment,Loan_Type
24,53,Full-time,167610.0,533,65607,36,8.47,3344,104402.0,3,No,Personal
86,52,Self-employed,33805.0,788,133538,48,14.12,1168,94898.0,5,No,Personal
4,72,Full-time,108950.0,612,48376,60,4.93,990,64234.0,4,Yes,Business
352,23,Unemployed,27110.0,752,66968,60,17.29,2622,91982.0,1,No,Auto
299,25,Full-time,185454.0,663,147031,36,16.25,699,9991.0,4,No,Business


In [77]:
X_train_processed = preprocessor.fit_transform(X_train)

X_test_processed = preprocessor.transform(X_test)

In [78]:
print("Before preprocessing:")
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)

print("\nAfter preprocessing:")
print("X_train_processed:", X_train_processed.shape)
print("X_test_processed:", X_test_processed.shape)

Before preprocessing:
X_train: (304, 12)
X_test: (76, 12)

After preprocessing:
X_train_processed: (304, 20)
X_test_processed: (76, 20)


In [79]:
from sklearn.linear_model import LogisticRegression

baseline_model = LogisticRegression(max_iter=1000)

baseline_model.fit(X_train_processed, y_train)

LogisticRegression(max_iter=1000)

In [80]:
y_pred = baseline_model.predict(X_test_processed)

print("Predicted values:")
print(y_pred)

Predicted values:
[0 0 1 0 0 0 0 0 0 1 0 1 0 0 1 0 1 0 0 0 0 1 1 1 0 1 0 1 0 0 0 0 1 1 1 1 1
 1 1 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 0 1 0 1 0 1 1 0 0 0 0 1 0 1
 1 0]


In [81]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import classification_report

In [82]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.7368421052631579


In [83]:
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[37 10]
 [10 19]]


In [84]:
import pandas as pd

cm_table = pd.DataFrame(
cm,
index=["Actual No", "Actual Yes"],
columns=["Predicted No", "Predicted Yes"]
)

cm_table

,Predicted No,Predicted Yes
Actual No,37,10
Actual Yes,10,19


In [85]:
precision = precision_score(y_test, y_pred, zero_division=0)

recall = recall_score(y_test, y_pred, zero_division=0)

f1 = f1_score(y_test, y_pred, zero_division=0)

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.6551724137931034
Recall: 0.6551724137931034
F1-score: 0.6551724137931034


In [86]:
print("Classification Report:")
print(classification_report(y_test, y_pred, target_names=["No", "Yes"], zero_division=0))

Classification Report:
              precision    recall  f1-score   support

          No       0.79      0.79      0.79        47
         Yes       0.66      0.66      0.66        29

    accuracy                           0.74        76
   macro avg       0.72      0.72      0.72        76
weighted avg       0.74      0.74      0.74        76



In [87]:
print("Model Evaluation Summary")

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

print("\nInterpretation:")
print("The baseline model gives us an initial performance result.")
print("We can use this result as a benchmark for comparing more advanced models later.")

Model Evaluation Summary
Accuracy: 0.7368421052631579
Precision: 0.6551724137931034
Recall: 0.6551724137931034
F1-score: 0.6551724137931034

Interpretation:
The baseline model gives us an initial performance result.
We can use this result as a benchmark for comparing more advanced models later.


In [88]:
print("Baseline Model Results:")
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Baseline Model Results:
Accuracy: 0.7368421052631579
Precision: 0.6551724137931034
Recall: 0.6551724137931034
F1-score: 0.6551724137931034


In [89]:
print("Interpretation:")
print("The baseline model provides an initial performance benchmark.")
print("Accuracy shows how many predictions were correct overall.")
print("Precision shows how reliable the positive early repayment predictions were.")
print("Recall shows how many actual early repayment customers the model was able to identify.")
print("F1-score balances precision and recall.")

Interpretation:
The baseline model provides an initial performance benchmark.
Accuracy shows how many predictions were correct overall.
Precision shows how reliable the positive early repayment predictions were.
Recall shows how many actual early repayment customers the model was able to identify.
F1-score balances precision and recall.


In [90]:
print("Business Interpretation:")
print("This model can help a financial institution identify customers who may be more likely to repay their loans early.")
print("The financial institution could use this information to support loan planning, cash flow management, and customer strategy.")
print("However, the model should be used carefully and should not be the only basis for financial decisions.")

Business Interpretation:
This model can help a financial institution identify customers who may be more likely to repay their loans early.
The financial institution could use this information to support loan planning, cash flow management, and customer strategy.
However, the model should be used carefully and should not be the only basis for financial decisions.


In [91]:
print("Model Limitations:")
print("1. This is only a baseline Logistic Regression model.")
print("2. The model may not capture complex patterns in customer financial behaviour.")
print("3. Some important customer information may be missing from the dataset.")
print("4. The model can make incorrect predictions.")
print("5. More testing and comparison with other models are needed before real business use.")

Model Limitations:
1. This is only a baseline Logistic Regression model.
2. The model may not capture complex patterns in customer financial behaviour.
3. Some important customer information may be missing from the dataset.
4. The model can make incorrect predictions.
5. More testing and comparison with other models are needed before real business use.


Assignment Explanation:

1. What is the model trying to predict?
The model is trying to predict whether a customer will repay their loan early or not.

2. What dataset was used?
The dataset used in this assignment is the finance early loan repayment prediction dataset.

3. What features and target were selected?
The selected features include customer, loan, income, credit, payment, and savings information. The target variable selected is `Early_Repayment`.

4. What result was obtained?
The Logistic Regression baseline model was trained and evaluated using accuracy, precision, recall, F1-score, confusion matrix, and classification report:

   Evaluation Results:
- Accuracy: 0.7368
- Precision: 0.6552
- Recall: 0.6552
- F1-score: 0.6552

5. One limitation of the model or dataset?
One limitation of this project is that the dataset may not fully represent real-world financial behavior. In addition, some important financial factors that affect early loan repayment may not be included in the dataset.
